(DS-536) refactor!: remove distro defaults settings to add from strain/conf file

JuanDavidBuitrago opened the pull request:
## Description

This PR remove ``Distro defaults settings`` to have a Tutor plugins flexible for different users, The principal idea is that people use Distro but with the packages that want to use, making the settings in a config/strain file.

For example, our Olmo master version is set in strain.yml file and contain ``Django plugins``/``Themes`` that use our SAAS.
Here are the settings that were removed and added in strain file: https://github.com/eduNEXT/ednx-strains/pull/46/files

### How to test

#### Using stack builder in your local to build the image
1. Make a strain.yml in a new work directory, use this [strain.yml](https://github.com/eduNEXT/ednx-strains/blob/jdb/add_default_settings_to_distro/olmo/master/strain.yml)
2. Inside the work directory, run ``stack strain create`` command
3. Next run ``stack strain local configure -s install_plugins -s enable_plugins -s save_config -s extra_commands``
4. After this ``run stack strain local build openedx --no-cache``
5. You can see the image was build correctly!

#### Using Picasso to build the image
1. Go to [Picasso-strain-builder](https://picasso.dedalo.edunext.co/job/Picasso-strain-builder/build?delay=0sec)
2. In ``STRAIN_REPOSITORY_BRANCH`` set ``jdb/add_default_settings_to_distro``, it's the branch where are the settings to Distro
3. In `` STRAIN_PATH `` set ``olmo/master``, it's the strain that has the Distro settings
4. Click on ``Build`` button
5. You can see the image was build correctly!

Alec4r commented:
@JuanDavidBuitrago this looks good for me, just a comment, for next PRs, please split the CI and new feature in differents PRs and I have a question, are there other strains using olive or olmo?

JuanDavidBuitrago commented:
> @JuanDavidBuitrago this looks good for me, just a comment, for next PRs, please split the CI and new feature in differents PRs and I have a question, are there other strains using olive or olmo?

@Alec4r I split the CI commit, is in this [PR](https://github.com/eduNEXT/tutor-contrib-edunext-distro/pull/47).
There are other strains, you can see [here](https://github.com/eduNEXT/ednx-strains/pull/46/files), but this all have the settings to continue working as expected.

> Also, what will happend if my config.yml doesn't have DISTRO_EXTRA_MIDDLEWARES or the others old default variables? will it work?

I fixed all problems when the defaults variables aren't defined, so if a user doesn't need to use all of them, Distro continue working normally

Can you check again?, please

Alec4r commented on tutordistro/patches/openedx-dockerfile-post-python-requirements:
```diff
@@ -1,4 +1,4 @@
-{%- if DISTRO_INSTALL_EDNX_REQUIREMENTS %}
+{%- if DISTRO_INSTALL_EDNX_REQUIREMENTS is defined and DISTRO_INSTALL_EDNX_REQUIREMENTS %}
```
Maybe we could remove this and add a setting like 


```yaml
INSTALL_EXTRA_FILE_REQUIREMENTS:
    - ./requirements/edunext/base.txt
    - ./requirements/nelp/test.txt
```

This is more generic maybe? Remember that these changes are for remove all "mandatory distro defaults"

JuanDavidBuitrago replied:
Yes, you're right. I will add the setting.

Alec4r commented on tutordistro/patches/cms-env:
```diff
@@ -1,2 +1,4 @@
 USE_EOX_TENANT: True
```
Maybe we need to create a feature to add settings to these files, because we need to remove all the EOX_ settings too.

JuanDavidBuitrago replied:
Ok, I will check this cases

dcoa commented:
We should update the documentation to remove all references to default values.

JuanDavidBuitrago commented:
Hi @Alec4r 

Could you check the new changes please?

MaferMazu commented on tutordistro/patches/openedx-lms-production-settings:
```diff
@@ -1,26 +1,32 @@
 
+{%- for pkg in iter_values_named(suffix="_DPKG") %}
+{%- if pkg["name"] == 'eox-theming' %}
 ################## EOX_THEMING ##################
```
Leaving this is not as leave the default? If someone wants to install eox-theming if not their responsibility to add this config somewhere?

JuanDavidBuitrago replied:
If user want to use eox-theming this is the basic configuration, and we have some cases at the moment:
1. We could remove this setting and leave the responsibility to users, but remember that we own both products, it's our responsibility to these work as expected. 
2. We could add some additional feature to add this setting in config.yml, but as I see it, each time this file becomes heavier.
3. With this configuration we make sure that eox-theming is used well if the user wants to use it, otherwise the configuration is simply not added, and in case the plugin has many changes that break the current behavior, we could deprecate this configuration. In the Changelog, I have not seen changes that break the current functionality in the last few releases.

At this point I think we can continue to work with option three, it is not a change that affects the goal of removing the default settings, because it is simply a consequence of using eox-theming and it is optional.

MaferMazu commented on README.md:
```diff
@@ -117,6 +118,8 @@ You can use the same steps that in **How to add a new package** just set the var
 - DISTRO_EOX_AUDIT_MODEL_DPKG
 - DISTRO_EOX_TAGGING_DPKG
 
+:warning: **NOTE**: From Olmo version Distro has not defaulted packages. Now it is necessary to add the packages you want in ``config.yml`` file. See [How to add a new package](./docs/how_to_add_new_packages.rst)
```
The default packages section doesn't have sense with his changes. I suggest dropping that section. What do you think?
The same with the override and disable packages.

JuanDavidBuitrago replied:
You are right on this point, but I do not consider it prudent to remove this content, because we have many users using the Distro with other versions of Tutor. It is better to give some time for users to start using the Distro for this new purpose and we might update the documentation in the future.

MaferMazu commented on README.md:
```diff
@@ -129,11 +132,15 @@ Development environment take this changes with _tutor config save_ and restart,
 Declare the path of your themes using `tutor config save --set DISTRO_THEMES_ROOT="your_path"`,
 by default the themes path goes here **/openedx/themes**
 
+:warning: **NOTE**: From Olmo version Distro has not defaulted themes path. Now it is necessary to add the themes path in ``config.yml`` file or running command above.
```
The same with this section of default themes. If we do not have default themes, we should drop this section. A good piece of info to put here is How to add a theme (and if you want, you can use a Bragi as example)

JuanDavidBuitrago replied:
Same comment here.

> You are right on this point, but I do not consider it prudent to remove this content, because we have many users using the Distro with other versions of Tutor. It is better to give some time for users to start using the Distro for this new purpose and we might update the documentation in the future.

MaferMazu commented on README.md:
```diff
@@ -223,6 +234,47 @@ DISTRO_EXTRA_MIDDLEWARES:
+    USE_EOX_TENANT: true
+  ]
+  lms_env: [
+    USE_EOX_TENANT: true,
```
I think it will be helpful to add an example of how to add a FEATURE setting because we use allow of that setting.

JuanDavidBuitrago replied:
I don't understand your comment here. As explained in the section, this would be the example of how to use the settings. When I mean that the list can grow, it does not depend on the Distro as such, but on what the user needs in the configurations, because the variables are not burned in the code.

MaferMazu commented on README.md:
```diff
@@ -223,6 +234,47 @@ DISTRO_EXTRA_MIDDLEWARES:
 
+## How to add extra files requirements
+
+You should set the variable **INSTALL_EXTRA_FILE_REQUIREMENTS** in your config.yml file if you need to install extra files with. The structure should be like:
```
The first time I read this, I didn't understand why I needed this. I read where you use it but I am still unsure how to use it. This is related to the command `tutor distro enable-private-packages`? Do I need this configuration to use that command?
It probably is a good idea to explain what is an extra file requirement, how the content looks, and if we need something extra; for example, if I decide to add one random file to a file directory in that, I think I need to have that file in a specific place.
I feel this part needs more info because it is not so easy to understand why I need this and what I need to do in order to do this work.

JuanDavidBuitrago replied:
It is just to install requirements files like: base.txt, coverage.txt, development.txt, doc.txt, testing.txt, ...
It depends on what the user needs at that moment.

MaferMazu reviewed (commented):
Sorry for the late review and many comments; this PR is a big breaking change, and I believe it is so important that the users of Distro understand what is happening.
I would like you to change a little bit the description of what the Distro is and what this plugin does after this PR and try to adapt the Readme to be easy to know what the user can do; for that reason, I let you several comments around to drop the default explanation and go straight to how to do the things they need to do.

I have a little question, when we merge this, are we going to have an "Olmo" version of the Distro with this plugin in v16? When we migrate a palm, are we planning to stay in v16?

The rest looks good.

JuanDavidBuitrago commented on tutordistro/patches/openedx-cms-production-settings:
```diff
@@ -1,5 +1,5 @@
 {% for pkg in iter_values_named(suffix="_DPKG") %}
-{%- if pkg != 'None' -%}
+{%- if pkg != 'None' and pkg["variables"] is defined %}
```
It is not the same, It's just to use Whitespace Control. To learn more of Jinja Whitespace Control, see https://jinja.palletsprojects.com/en/3.0.x/templates/#whitespace-control

MaferMazu replied:
Thanks for this. I didn't know.

JuanDavidBuitrago commented:
> Sorry for the late review and many comments; this PR is a big breaking change, and I believe it is so important that the users of Distro understand what is happening. I would like you to change a little bit the description of what the Distro is and what this plugin does after this PR and try to adapt the Readme to be easy to know what the user can do; for that reason, I let you several comments around to drop the default explanation and go straight to how to do the things they need to do.
> 
> I have a little question, when we merge this, are we going to have an "Olmo" version of the Distro with this plugin in v16? When we migrate a palm, are we planning to stay in v16?
> 
> The rest looks good.

I don't think the changes that were made drastically change the description of the Distro. It's just that now we're not going to have Django plugins and Themes by default, now users are free to use whatever they want in their environment.

Regarding the version, this continues to lead to the nomenclature according to the version of Tutor, as it has been working up to now. Remember that in Nutmeg (v14) we went from a v3.2.0 Tag to v14.2.0 just to be consistent with Tutor versions.

MaferMazu commented:
Yes, I really don't understand.

- If we are going to follow the Tutor version this change will create the v15.2.0 being a breaking change? (My proposal around this is to start this breaking change for Palm, and update the release to v16)
- How is not a default having code about eox-theming? (Probably my concern here is because I don't understand what will be the purpose of this plugin)
- In the readme said:
> Distro is an opinioned openedx distribution with some custom stuff to have an easy-to-use and a ready to deploy in local or in development openedx distribution. This can be watch like a tutor-plugin but is taken a little bit far away.
And If I see Distro as distribution, yes. But this plugin with these changes won't be the opinionated openedx distribution, maybe a tool we use to make easier our work (we are going to have our tools), but not the opinionated distribution. (Probably add some information more accurate about what this plugin really will do)
- And I do not agree with maintaining old documentation that is not consistent with the real behavior. (Adding comments for future behavior or past behavior is okay, but having old documentation with the current behavior in comments ... is not intuitive)

Maybe I need to understand better what we want to do with this plugin, for that reason, I am not feeling comfortable giving a review.

JuanDavidBuitrago commented:
> Yes, I really don't understand.
> 
>     * If we are going to follow the Tutor version this change will create the v15.2.0 being a breaking change? (My proposal around this is to start this breaking change for Palm, and update the release to v16)
> 
>     * How is not a default having code about eox-theming? (Probably my concern here is because I don't understand what will be the purpose of this plugin)
> 
>     * In the readme said:
> 
> 
> > Distro is an opinioned openedx distribution with some custom stuff to have an easy-to-use and a ready to deploy in local or in development openedx distribution. This can be watch like a tutor-plugin but is taken a little bit far away.
> 
> And If I see Distro as distribution, yes. But this plugin with these changes won't be the opinionated openedx distribution, maybe a tool we use to make easier our work (we are going to have our tools), but not the opinionated distribution. (Probably is a good idea to add some information more accurate about what this plugin really will do)
> 
>     * And I do not agree with maintaining old documentation that is not consistent with the real behavior. (Adding comments for future behavior or past behavior is okay, but having old documentation with the current behavior in comments ... is not intuitive)
> 
> 
> Maybe I need to understand better what we want to do with this plugin, for that reason, I am not feeling comfortable giving a review.

Hi @Alec4r, could you give me a hand here? Please

Alec4r commented:
> Yes, I really don't understand.
> 
> * If we are going to follow the Tutor version this change will create the v15.2.0 being a breaking change? (My proposal around this is to start this breaking change for Palm, and update the release to v16)
> * How is not a default having code about eox-theming? (Probably my concern here is because I don't understand what will be the purpose of this plugin)
> * In the readme said:
> 
> > Distro is an opinioned openedx distribution with some custom stuff to have an easy-to-use and a ready to deploy in local or in development openedx distribution. This can be watch like a tutor-plugin but is taken a little bit far away.
> 
> And If I see Distro as distribution, yes. But this plugin with these changes won't be the opinionated openedx distribution, maybe a tool we use to make easier our work (we are going to have our tools), but not the opinionated distribution. (Probably is a good idea to add some information more accurate about what this plugin really will do)
> 
> * And I do not agree with maintaining old documentation that is not consistent with the real behavior. (Adding comments for future behavior or past behavior is okay, but having old documentation with the current behavior in comments ... is not intuitive)
> 
> Maybe I need to understand better what we want to do with this plugin, for that reason, I am not feeling comfortable giving a review.

We have already talk about this, but I could explain again:

**Distro is an opinioned openedx distribution.**: Yes, The most important part is "is an opinioned openedx distribution", distro is just a number of requirements, settings, xblocks, that edunext considerate are the best, but is just an opinion, so we will move that to an one pager, documentation and strain, due to it's just an opinion, it's similar to NELP opinion, Pearson Opinion, are just opinions.

**with some custom stuff to have an easy-to-use and a ready to deploy in local or in development openedx distribution**: Second part, how to use this, ok, the answer for this part is "Distro plugin/Tutor Stack Plugin" that help us to load opinions (Strains) easier, but it's just another way to do it, because you can apply NELP opinion without use a plugin, you could use inline plugins, and you can configure each theme or private requirements without need "Distro plugin/Tutor Stack Plugin", but maybe with this plugin is easier than replicate each step.

Why on this version and not wait for palm?, because we need to test now, and be safe all is working while we can change between versions and not in palm while we are in the middle of a big migration.

If you want to see this like "Distro Plugin v2.0" it's ok, this is a big rework of our old plugin.

And of course this will need new documentation, I totally agree with you.

Now, this is the reason why we used to have difference version and we couldn't use just "Tutor Semantic Version", because our break change doesn't depends of tutor version, this is a tool more than a tutor plugin, we are just using tutor technologies for make easier our work.

Alec4r commented:
@JuanDavidBuitrago did you already do some test using picasso? could you give me the link or number of the build?

JuanDavidBuitrago commented:
> @JuanDavidBuitrago did you already do some test using picasso? could you give me the link or number of the build?

Yes @Alec4r, I did the build with this [olmo strain](https://github.com/eduNEXT/ednx-strains/blob/jdb/add_default_settings_to_distro/olmo/master/strain.yml) and the job in Picasso is [#121](https://picasso.dedalo.edunext.co/job/Picasso-strain-builder/121/)

Let me know if you need anything else.
